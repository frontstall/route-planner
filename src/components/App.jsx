import React from 'react';
import uniqueId from 'lodash/uniqueId';
import pickBy from 'lodash/pickBy';
import isEqual from 'lodash/isEqual';
import { DragDropContext } from 'react-beautiful-dnd';

import Button from './Button';
import PointsList from './PointsList';
import Map from './Map';
import Form from './Form';

import getYandexMapFacade from '../facades/yandexMapFacade';
import { DEFAULT_MAP_CENTER } from '../tools/const';

class App extends React.Component {
  constructor(props) {
    super(props);

    const { points, pointsIds } = this.props;
    this.state = {
      points,
      pointsIds,
      addingNewPoint: false,
      center: DEFAULT_MAP_CENTER,
      mapLoaded: false,
      map: null,
    };
  }

  static defaultProps = {
    points: {},
    pointsIds: [],
  };

  componentDidUpdate(prevProps, prevState) {
    if (this.shouldRenderPoints(prevState)) {
      const mapFacade = this.getMapFacade();

      mapFacade.clearMap();
      mapFacade.renderPoints();
      mapFacade.renderRoute();
    }
  }

  openNewPointForm = () => {
    this.setState(() => ({ addingNewPoint: true }));
  };

  closeNewPointForm = () => {
    this.setState(() => ({ addingNewPoint: false }));
  };

  addPoint = newPointName => {
    const mapFacade = this.getMapFacade();
    const id = uniqueId();
    const center = mapFacade.getCenter();
    const newPointData = { id, name: newPointName, coordinates: center };

    this.setState(({ points, pointsIds }) => ({
      points: {
        ...points,
        [id]: newPointData,
      },
      pointsIds: [...pointsIds, id],
    }));
  };

  removePoint = pointId => {
    this.setState(({ points, pointsIds }) => ({
      points: pickBy(points, ({ id }) => id !== pointId),
      pointsIds: pointsIds.filter(id => id !== pointId),
    }));
  };

  swapListItems = ({ destination, draggableId, source }) => {
    if (!destination) {
      return;
    }

    const { index: sourceIndex } = source;
    const { index: destinationIndex } = destination;

    if (destinationIndex === sourceIndex) {
      return;
    }

    this.setState(({ pointsIds }) => {
      const updatedPointsIds = pointsIds.slice();

      updatedPointsIds.splice(sourceIndex, 1);
      updatedPointsIds.splice(destinationIndex, 0, draggableId);

      return { pointsIds: updatedPointsIds };
    });
  };

  onYandexMapInitialScriptLoad = () => {
    const { ymaps } = window;
    ymaps.ready(this.initMap);
  };

  initMap = () => {
    const { center, zoom = 15 } = this.state;
    const { ymaps } = window;
    const map = new ymaps.Map('map', { center, zoom });
    this.setState(() => ({
      mapLoaded: true,
      map,
    }));
  };

  getMapFacade = () => {
    const { map, points, pointsIds } = this.state;

    return getYandexMapFacade(map, points, pointsIds, this.onPointDragEnd);
  };

  onPointDragEnd = id => evt => {
    const movedPoint = evt.get('target');
    const movedPointCoordinates = movedPoint.geometry.getCoordinates();

    this.setState(({ points }) => ({
      points: {
        ...points,
        [id]: { ...points[id], coordinates: movedPointCoordinates },
      },
    }));
  };

  shouldRenderPoints = prevState => {
    const { points: prevPoints, pointsIds: prevPointsIds } = prevState;
    const { points, pointsIds } = this.state;

    return !isEqual(prevPoints, points) || !isEqual(prevPointsIds, pointsIds);
  };

  render() {
    const { points, addingNewPoint, pointsIds, mapLoaded } = this.state;

    return (
      <DragDropContext onDragEnd={this.swapListItems}>
        <div className="app">
          <header className="app__header header">Funbox test app</header>
          <div className="app__container">
            <div className="app__points points">
              {!addingNewPoint && (
                <Button
                  onClick={this.openNewPointForm}
                  text="Новая точка маршрута"
                  disabled={!mapLoaded || addingNewPoint}
                  className="points__add-point-button button--big"
                />
              )}

              {addingNewPoint && (
                <Form
                  onSubmit={this.addPoint}
                  onClose={this.closeNewPointForm}
                />
              )}
              {pointsIds.length > 0 && (
                <PointsList
                  points={points}
                  pointsIds={pointsIds}
                  onPointRemove={this.removePoint}
                />
              )}
            </div>
            <Map
              onMapApiInitialScriptLoad={this.onYandexMapInitialScriptLoad}
              className="app__map"
            />
          </div>
        </div>
      </DragDropContext>
    );
  }
}

export default App;
