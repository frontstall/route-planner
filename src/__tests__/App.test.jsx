import React from 'react';
import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import App from '../components/App';

const getMapFacadeMocked = () => ({
  clearMap: jest.fn(),
  renderRoute: jest.fn(),
  renderPoints: jest.fn(),
  getCenter: jest.fn(),
  getAddress: jest.fn(),
});

jest.mock('react-beautiful-dnd', () => ({
  Droppable: ({ children }) =>
    children(
      {
        draggableProps: {
          style: {},
        },
        innerRef: jest.fn(),
      },
      {},
    ),
  Draggable: ({ children }) =>
    children(
      {
        draggableProps: {
          style: {},
        },
        innerRef: jest.fn(),
      },
      {},
    ),
  DragDropContext: ({ children }) => children,
}));

const points = {
  test1: {
    coordinates: [55.759, 37.64],
    id: 'test1',
    name: 'point one',
  },
  test2: {
    coordinates: [55.7621, 37.6429],
    id: 'test2',
    name: 'point two',
  },
  test3: {
    coordinates: [55.76029, 37.6482],
    id: 'test3',
    name: 'point three',
  },
  test4: {
    coordinates: [55.75556, 37.6488],
    id: 'test4',
    name:
      'point four with really long name you cant believe it can be sooo long',
  },
};

const pointsIds = ['test1', 'test2', 'test3', 'test4'];

describe('renders App', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders correctly', () => {
    const wrapper = mount(<App points={points} pointsIds={pointsIds} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

describe('adds point', () => {
  const wrapper = mount(<App />);
  wrapper.setState({ mapLoaded: true });
  wrapper.instance().getMapFacade = getMapFacadeMocked;

  it('points list should be empty', () => {
    const pointsList = wrapper.find('.points-list__item');

    expect(pointsList.length).toEqual(0);
  });

  it('should be button for opening form', () => {
    const button = wrapper.find('Button.points__add-point-button');

    expect(button.length).toEqual(1);
  });

  it('button should open form', () => {
    const button = wrapper.find('Button.points__add-point-button');

    button.simulate('click');

    const form = wrapper.find('.form');

    expect(form.length).toEqual(1);
  });

  it('should be button to add point', () => {
    const button = wrapper.find('Button.button--iconed-plus');

    expect(button.length).toEqual(1);
  });

  it('form submitting should add one point', () => {
    const form = wrapper.find('.form');

    form.simulate('submit');

    const pointsList = wrapper.find('li.points-list__item');
    expect(pointsList.length).toEqual(1);
  });
});

describe('remove point', () => {
  const wrapper = mount(<App points={points} pointsIds={pointsIds} />);
  wrapper.instance().getMapFacade = getMapFacadeMocked;

  it('every point item should have button for removing', () => {
    const buttons = wrapper.find('Button.point__remove-button');

    expect(pointsIds.length).toEqual(buttons.length);
  });

  it('click on button should remove one point', () => {
    const pointsCountBeforeDeletion = wrapper.find('li.points-list__item')
      .length;
    const button = wrapper.find('Button.point__remove-button').first();

    button.simulate('click');

    const pointsCountAfterDeletion = wrapper.find('li.points-list__item')
      .length;
    expect(pointsCountBeforeDeletion - pointsCountAfterDeletion).toEqual(1);
  });
});
