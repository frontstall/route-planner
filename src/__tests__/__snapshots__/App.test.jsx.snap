// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders App renders correctly 1`] = `
<App
  points={
    Object {
      "test1": Object {
        "coordinates": Array [
          55.759,
          37.64,
        ],
        "id": "test1",
        "name": "point one",
      },
      "test2": Object {
        "coordinates": Array [
          55.7621,
          37.6429,
        ],
        "id": "test2",
        "name": "point two",
      },
      "test3": Object {
        "coordinates": Array [
          55.76029,
          37.6482,
        ],
        "id": "test3",
        "name": "point three",
      },
      "test4": Object {
        "coordinates": Array [
          55.75556,
          37.6488,
        ],
        "id": "test4",
        "name": "point four with really long name you cant believe it can be sooo long",
      },
    }
  }
  pointsIds={
    Array [
      "test1",
      "test2",
      "test3",
      "test4",
    ]
  }
>
  <DragDropContext
    onDragEnd={[Function]}
  >
    <div
      className="app"
    >
      <header
        className="app__header header"
      >
        Route planner
      </header>
      <div
        className="app__container"
      >
        <div
          className="app__points points"
        >
          <Button
            className="points__add-point-button button--big"
            disabled={true}
            onClick={[Function]}
            text="Новая точка маршрута"
          >
            <button
              className="button points__add-point-button button--big"
              disabled={true}
              onClick={[Function]}
              type="button"
            >
              Новая точка маршрута
            </button>
          </Button>
          <PointsList
            onPointRemove={[Function]}
            points={
              Object {
                "test1": Object {
                  "coordinates": Array [
                    55.759,
                    37.64,
                  ],
                  "id": "test1",
                  "name": "point one",
                },
                "test2": Object {
                  "coordinates": Array [
                    55.7621,
                    37.6429,
                  ],
                  "id": "test2",
                  "name": "point two",
                },
                "test3": Object {
                  "coordinates": Array [
                    55.76029,
                    37.6482,
                  ],
                  "id": "test3",
                  "name": "point three",
                },
                "test4": Object {
                  "coordinates": Array [
                    55.75556,
                    37.6488,
                  ],
                  "id": "test4",
                  "name": "point four with really long name you cant believe it can be sooo long",
                },
              }
            }
            pointsIds={
              Array [
                "test1",
                "test2",
                "test3",
                "test4",
              ]
            }
          >
            <Droppable
              droppableId="points-list"
            >
              <ul
                className="points-list"
              >
                <Point
                  className="points-list__item"
                  id="test1"
                  index={0}
                  key="test1"
                  name="point one"
                  onRemove={[Function]}
                >
                  <Draggable
                    draggableId="test1"
                    index={0}
                  >
                    <li
                      className="point points-list__item"
                      style={Object {}}
                    >
                      <span
                        className="point__name"
                        title="point one"
                      >
                        point one
                      </span>
                      <span
                        className="point__address"
                      />
                      <Button
                        className="point__remove-button button--iconed-cross"
                        onClick={[Function]}
                        title="Удалить"
                      >
                        <button
                          className="button point__remove-button button--iconed-cross"
                          disabled={false}
                          onClick={[Function]}
                          title="Удалить"
                          type="button"
                        />
                      </Button>
                    </li>
                  </Draggable>
                </Point>
                <Point
                  className="points-list__item"
                  id="test2"
                  index={1}
                  key="test2"
                  name="point two"
                  onRemove={[Function]}
                >
                  <Draggable
                    draggableId="test2"
                    index={1}
                  >
                    <li
                      className="point points-list__item"
                      style={Object {}}
                    >
                      <span
                        className="point__name"
                        title="point two"
                      >
                        point two
                      </span>
                      <span
                        className="point__address"
                      />
                      <Button
                        className="point__remove-button button--iconed-cross"
                        onClick={[Function]}
                        title="Удалить"
                      >
                        <button
                          className="button point__remove-button button--iconed-cross"
                          disabled={false}
                          onClick={[Function]}
                          title="Удалить"
                          type="button"
                        />
                      </Button>
                    </li>
                  </Draggable>
                </Point>
                <Point
                  className="points-list__item"
                  id="test3"
                  index={2}
                  key="test3"
                  name="point three"
                  onRemove={[Function]}
                >
                  <Draggable
                    draggableId="test3"
                    index={2}
                  >
                    <li
                      className="point points-list__item"
                      style={Object {}}
                    >
                      <span
                        className="point__name"
                        title="point three"
                      >
                        point three
                      </span>
                      <span
                        className="point__address"
                      />
                      <Button
                        className="point__remove-button button--iconed-cross"
                        onClick={[Function]}
                        title="Удалить"
                      >
                        <button
                          className="button point__remove-button button--iconed-cross"
                          disabled={false}
                          onClick={[Function]}
                          title="Удалить"
                          type="button"
                        />
                      </Button>
                    </li>
                  </Draggable>
                </Point>
                <Point
                  className="points-list__item"
                  id="test4"
                  index={3}
                  key="test4"
                  name="point four with really long name you cant believe it can be sooo long"
                  onRemove={[Function]}
                >
                  <Draggable
                    draggableId="test4"
                    index={3}
                  >
                    <li
                      className="point points-list__item"
                      style={Object {}}
                    >
                      <span
                        className="point__name"
                        title="point four with really long name you cant believe it can be sooo long"
                      >
                        point four with really long name you cant believe it can be sooo long
                      </span>
                      <span
                        className="point__address"
                      />
                      <Button
                        className="point__remove-button button--iconed-cross"
                        onClick={[Function]}
                        title="Удалить"
                      >
                        <button
                          className="button point__remove-button button--iconed-cross"
                          disabled={false}
                          onClick={[Function]}
                          title="Удалить"
                          type="button"
                        />
                      </Button>
                    </li>
                  </Draggable>
                </Point>
              </ul>
            </Droppable>
          </PointsList>
        </div>
        <Map
          className="app__map"
          onMapApiInitialScriptLoad={[Function]}
        >
          <div
            className="map app__map"
            id="map"
          />
        </Map>
      </div>
    </div>
  </DragDropContext>
</App>
`;
